### PR TITLE
Feature/quiz

### PR DESCRIPTION
**Pull request by aserhany**

### CODE REVIEW

File: app/src/main/res/layout/card_test.xml [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (kordianbruck) [inline app/src/main/res/layout/card_test.xml]: Please remove this file, it is has been moved to "card_survey"

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -472,7 +472,7 @@ Signature: %5$s</string>
     <string name="complete_question_form">Please complete question form</string>
     <string name="select_one_faculty">At least select one faculty</string>
     <string name="question_deleted">Question has been deleted</string>
-    <string name="choose_own_faculty" translatable="false">Please choose your faculty:</string>
+    <string name="choose_own_faculty" translatable="false">Please choose your major:</string>
```
Comment (kordianbruck) [inline app/src/main/res/values/strings.xml]: This was already fixed by me, please revert this change.

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -101,7 +101,7 @@
     <!-- Startup Wizard -->
     <string name="token_not_enabled">TUMOnline-Access-Token is not enabled</string>
     <string name="startup_wizard_uppercase">STARTUP-WIZARD</string>
-    <string name="welcome_msg"><b>Welcome to TUM Campus App!</b>\n\nIn order to setup the TUMOnline access, you need to provide your <b>LRZ Identification</b>.\n\nThis Id has the following format:\n<i>xx00xxx</i>\n(2 letters, 2 digits, 3 letters)</string>
+    <string name="welcome_msg"><b>Welcome to TUM Campus App!</b>\n\nIn order to setup the TUMOnline access, you need to provide your <b>LRZ Identification</b> and your <b>major of studies</b>.\n\nThis LRZ Id has the following format:\n<i>xx00xxx</i>\n(2 letters, 2 digits, 3 letters)</string>
```
Comment (kordianbruck) [inline app/src/main/res/values/strings.xml]: This String is wrong, please revert this change.

File: app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java [hunk 4]
```diff
@@ -340,12 +327,35 @@ void replaceIntoDBOpenQuestions(Question q) {
     }
 
     void replaceIntoDb(Faculty f) {
-        db.execSQL("REPLACE INTO faculties (faculty, name) VALUES (?, ?)", new String[]{f.getId(), f.getName()});
-    }
+        // Unfortunately I had to do it like that and not with a Replace Into statment because for some reason the replace statement doesn't work correctly
+        Cursor c = db.rawQuery("SELECT * FROM faculties WHERE faculty = ?",new String[]{f.getId()});
+        if(c.moveToFirst()){
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java]: Please optimize this "if". You don't need the try/catch block in both parts of the if.

File: app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java [hunk 5]
```diff
@@ -111,45 +112,31 @@ public void updateQuestion(Question question, String updateField) {
             db.endTransaction();
         }
 
-        //Tigger sync if we are connected currently
+        //Tiger sync if we are connected currently
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java]: Typo: Trigger

File: app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java [hunk 6]
```diff
@@ -26,22 +27,21 @@
...
-    private static int answerNo = 2;
-    private static int answerFlag = -1;
+
 
     public SurveyManager(Context context) {
         super(context);
 
         db.execSQL("CREATE TABLE IF NOT EXISTS surveyQuestions (id INTEGER PRIMARY KEY, question VARCHAR, yes BOOLEAN, no BOOLEAN, flagged BOOLEAN, answered BOOLEAN, synced BOOLEAN)");
         db.execSQL("CREATE TABLE IF NOT EXISTS faculties (faculty INTEGER, name VARCHAR)");
         db.execSQL("CREATE TABLE IF NOT EXISTS survey1 (id INTEGER PRIMARY KEY AUTOINCREMENT, date VARCHAR,userID VARCHAR, question TEXT, faculties TEXT, yes INTEGER,  no INTEGER, flags INTEGER)");
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/models/managers/SurveyManager.java]: Please remove unused tables.

File: app/src/main/java/de/tum/in/tumcampusapp/models/TUMCabeClient.java [hunk 7]
```diff
@@ -45,9 +45,10 @@
     private static final String API_LOCATIONS = "/locations/";
     private static final String API_DEVICE = "/device/";
     private static final String API_QUESTION = "/question/";
-    private static final String API_FACULTIES = "/faculty/";
     private static final String API_ANSWER_QUESTION = "/question/answer";
     private static final String API_OWN_QUESTIONS = "/question/my";
+    private static final String API_FACULTY = "/faculty";
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/models/TUMCabeClient.java]: Please add the trailing slash to all API endpoints.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 8]
```diff
@@ -99,8 +101,10 @@ public void onCreate(Bundle savedInstanceState) {
     }
 
     @Override
-    public void onRefresh() {}
-  //set up the respone tab layout dynamically depending on number of questions
+    public void onRefresh() {
+    }
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Please don't override methods if they are empty. Just remove these lines.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 9]
```diff
@@ -48,22 +47,26 @@
...
+    private Button submitSurveyButton, facultiesButton;
+    private ArrayList<String> questions = new ArrayList<>();
+    private ArrayList<String> selectedFaculties = new ArrayList<>();
+    private boolean[] checked = new boolean[14];
+    private LinearLayout mainResponseLayout, questionsLayout;
+    String chosenFaculties = "";
+    private ArrayList<String> fetchedFaculties = new ArrayList<>();
+    private ViewGroup parentView;
+    private String[] numQues = new String[3];
     private SurveyManager surveyManager;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Please check, if all of these variables really need to be class vars. Maybe you can move some of them to the functions?

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 10]
```diff
@@ -196,6 +200,29 @@ private void setUpResponseTab() {
 
     }
 
+    //Handels clicking on 'delete' button of an own question in responses tab
+    View.OnClickListener clicks = new View.OnClickListener() {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: The name of this OnClickListener variable is not good.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 11]
```diff
@@ -204,35 +231,35 @@ private void zoomOutanimation(View v) {
         parentView.startAnimation(zoomOut);
         zoomOut.setAnimationListener(new Animation.AnimationListener() {
             @Override
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Why do you need to set the AnimationListener here? Just remove it, as you are not animating anything? Or add a comment on why it is needed.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 12]
```diff
@@ -204,35 +231,35 @@ private void zoomOutanimation(View v) {
...
-    };
-    //get all views by IDs
+    /**
+     * For restarting SurveyAcitivity. Used upon reconnecting to internet after no connection, or when submitting question(s)
+     */
+    private void restartActivity() {
+        Intent i = getIntent();
+        i.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
+        finish();
+        startActivity(i);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: You should not call the activity from this context. It will break moving back to the previous activity using the back button.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 13]
```diff
@@ -440,21 +452,76 @@ protected void onPostExecute(Void v) {
...
-        startActivity(i);
+
+
+    /**
+     * Help function for plausibility check of the question form
+     *
+     * @return true if everything was entered correctly else snackbar for requesting to complete questions.
+     */
+    private boolean checkSurveyData() {
+        boolean done = true;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Instead of returning a boolean here, why not return the form data in an array or appropriate model? Using class variables increases the complexity and reduces readability of the code.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 14]
```diff
@@ -440,21 +452,76 @@ protected void onPostExecute(Void v) {
...
+            Snackbar.make(findViewById(R.id.drawer_layout), getResources().getString(R.string.complete_question_form), Snackbar.LENGTH_LONG).show();
+        }
+        return false;
+    }
+
+    /**
+     * Help function for clearing data and layout entries after submitting questions
+     */
+    private void clearData() {
+        selectedFaculties.clear();
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: If you wouldn't use class variables, this function would not be needed.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 15]
```diff
@@ -440,21 +452,76 @@ protected void onPostExecute(Void v) {
...
+    }
+
+    //get selected faculties from spinner
+
+    /**
+     * Help function for getting all faculties as a string from the spinner
+     * @param arrayList: selected faculties in spinner as arraylist
+     * @return: selected faculties as String
+     */
+    public String getSelectedFaculties(ArrayList<String> arrayList) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Can you generalize this method and put it into the Utils?

PHP has a very similar function: http://php.net/manual/de/function.implode.php
Bonus points if use something from here: http://stackoverflow.com/questions/11248119/java-equivalent-of-phps-implode-array-filter-array

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java [hunk 16]
```diff
@@ -470,22 +537,20 @@ private void setUpTabHost() {
...
-        return !et.getText().toString().isEmpty();
-    }
 
-    //check if user is allowed to submit survey depending on the last survey date and number of question he submitted before
+    /**
+     * check if user is allowed to submit survey depending on the last survey date and number of question he submitted in the last week
+     */
     @SuppressLint("SetTextI18n")
-    private void userAllowed() {
+    private void numberOfQuestionsUserAllowed() {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: This function should probably be in the manager?

Maybe decouple UI changes and checking if allowed from each other to make this function simpler.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 15 above)
Comment (aserhany) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: TextUtils.join(",", selectedFaculties) where selectedFaculties is ArrayList<String>

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 12 above)
Comment (aserhany) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: We have just tested the back button after refreshing and it works fine. Is there smth we are not taking into consideration while testing that? Many thanks for your reply in advance.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 9 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Done.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 10 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: adjusted to deleteQuestion.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 11 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: zoomOut is used upon deleting question.
Code is now commeneted.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 13 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: adjusted , it now returns arraylist of questions.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 14 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: clearData is needed for class vars.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 15 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: Added to Utils .
public static String arrayListToString(ArrayList<String> array) and
public static String arrayToString(String[] array)
for both String[] and ArrayList<String>

File: app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java (hunk 16 above)
Comment (mhmadayad) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/SurveyActivity.java]: merged with another function.
setUpSpinnerForQuestionsNumber()
//checks available number of questions and handles numberOfquestions spinner liseneter
